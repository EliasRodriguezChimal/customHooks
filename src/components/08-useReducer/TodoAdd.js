import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ''
    });

    const handleSubmit = (e) => {
        //Esta funcion se ejecuta en cuanto se presiona el boton del formulario ya
        //que usa la funcion handleSubmit.
        e.preventDefault();

        //Las siguientes lineas sirven para hacer una comprobacion y evitar que se agreguen
        //lineas vacias como items de la lista
        if (description.trim().length <= 1) {
            return;
        }

         //Creamos un nuevo TODO
         const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        //Disparamos la action mediante el dispatch con el reducer correspondiente
        handleAddTodo(newTodo);
        reset();
    }
    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form
                className="d-grid gap-3"
                onSubmit={handleSubmit}
            >
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
