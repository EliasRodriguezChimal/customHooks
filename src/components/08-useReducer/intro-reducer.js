const initialState = [{
    id: 1,
    todo: 'Comprar awa',
    done: false
}];

const todoReducer = (state = initialState, action) => {

    //Esta funcion SIEMPRE tiene que resolverse sin disparar efectos secundarios, es decir,
    //debe resolverse con el state y el action que ha recibido, sin usar funciones externas
    //ni nada mas.

    //Aqui es donde debe agregarse el nuevo TODO
    if(action?.type === 'add'){
        //Aqui NO usamos push, en vez de eso, retornamos el nuevo estado gracias al operador
        //spread y agregando el payload contenido en action.
        return [...state, action.payload];
    }

    return state;   //Siempre debe retornarse un estado, aun si no se hizo ninguna operacion.
}

let todos = todoReducer();

//NOTA: Nunca usar push() en React porque modifica o muta el objeto.

const newTodo = {
    id: 2,
    todo: 'Comprar panecito',
    done: false
}

const agregatTodoAction = {
    type: 'add',
    payload: newTodo
}

todos = todoReducer(todos, agregatTodoAction);

console.log(todos);