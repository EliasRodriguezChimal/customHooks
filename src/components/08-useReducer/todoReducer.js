export const todoReducer = (state = [], action) => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];

        case 'delete':
            return state.filter(todo => todo.id !== action.payload); 
            //filter() retorna un nuevo arreglo con los elementos que cumplan cierta condicion

        //Forma corta del toggle.
        case 'toggle':
            return state.map(todo => 
                (todo.id === action.payload)
                    ? {...todo, done: !todo.done}
                    : todo
            )

        //Forma larga de realizar el toggle.
        case 'toggle-old':
            return state.map(todo => {
                if(todo.id === action.payload){
                    return {
                        ...todo, 
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            });
    
        default:
            return state;
    }
}