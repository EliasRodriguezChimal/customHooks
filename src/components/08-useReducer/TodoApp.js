import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { TodosList } from './TodosList';
import { TodoAdd } from './TodoAdd';

import './styles.css';

//La funcion 'init' se llama dentro del reducer, y todo lo que retorne sera considerado el
//initialState del mismo reducer.

//En este caso, la funcion init incluye el almacenamiento de datos persistente gracias a 
//Local Storage, permitiendo asi que, aunque se recargue el navegador, no se pierdan los datos.
const init = () => {
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender React UwU',
    //     done: false
    // }];

    //Para recuperar los TODOs del Local Storage, usamos la instruccion inferior, la cual nos permite
    //convertir la string guardada previamente a un objeto, y en caso de que no exista nada (que el 
    //Local Storage este vacio), la funcion JSON.parse() devolvera 'null', y para eso se usa el OR ( || )
    //que se ve abajo, el cual nos permitira devolver un arreglo vacio en caso de que dicha instruccion haya
    //retornado null.
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    //El 'dispatch' es una funcion a la cual se le envia una accion y sabra
    //que reducer le corresponde y que debe redibujar.
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    //El siguiente efecto nos permite guardar los TODOs, cada vez que sean modificados, en el 
    //local storage para evitar que se pierdan al recargar el navegador.

    //NOTA: El local storage SOLAMENTE ALMACENA STRINGS, NO OBJETOS, por lo que tenemos que pasar los
    //objetos por la funcion JSON.stringify() para convertir el objeto en un json que sera un string.
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos])

    //Funcion para agregar un TODO a la lista
    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    //Funcion para eliminar un TODO de la lista
    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }

        dispatch(action);
    }

    //Funcion para cambiar el valor 'done' de un TODO de la lista
    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr />

            <div className="row">
                <div className="col-7">
                    <TodosList
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd handleAddTodo={handleAddTodo}/>
                </div>
            </div>
        </div>
    )
}
