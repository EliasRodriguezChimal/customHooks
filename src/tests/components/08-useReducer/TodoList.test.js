import React from 'react';
import { shallow } from 'enzyme';
import { TodosList } from '../../../components/08-useReducer/TodosList';
import { demoTodos } from '../../fixtures/demoTodos';

describe('<TodoList /> tests', () => {
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const wrapper = shallow(
        <TodosList
            todos={demoTodos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
        />
    )

    test('should be displayed correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should have 2 <TodoListItem />', () => { 
        expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);

        expect(wrapper.find('TodoListItem').at(0).prop('handleToggle')).toEqual(expect.any(Function));
    });
})