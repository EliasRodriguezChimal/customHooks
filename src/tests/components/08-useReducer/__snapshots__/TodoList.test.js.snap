// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> tests should be displayed correctly 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender Firebase",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender AWS",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
